packages/marshal/src/structure.js

```diff
@@ -0,0 +1,215 @@
+// @ts-check
+
+import { assert, details as X, q } from '@agoric/assert';
+import { isPrimitive } from './helpers/passStyleHelpers.js';
+import { passStyleOf, everyPassableChild } from './passStyleOf.js';
+
+const { is, fromEntries, getOwnPropertyNames } = Object;
+
+const { ownKeys } = Reflect;
+
+/**
+ * This is the equality comparison used by JavaScript's Map and Set
+ * abstractions, where NaN is the same as NaN and -0 is the same as
+ * 0. Marshal serializes -0 as zero, so the semantics of our distributed
+ * object system does not distinguish 0 from -0.
+ *
+ * `sameValueZero` is the EcmaScript spec name for this equality comparison,
+ * but TODO we need a better name for the API.
+ *
+ * @param {any} x
+ * @param {any} y
+ * @returns {boolean}
+ */
+export const sameValueZero = (x, y) => x === y || is(x, y);
+harden(sameValueZero);
+
+/**
+ * TODO If the path to the non-structure becomes an important diagnostic,
+ * consider factoring this into a checkStructure that also takes
+ * a `path` and a `check` function.
+ *
+ * @param {Passable} passable
+ * @returns {boolean}
+ */
+const isStructureInternal = passable => {
+  const passStyle = passStyleOf(passable);
+  switch (passStyle) {
+    case 'null':
+    case 'undefined':
+    case 'string':
+    case 'boolean':
+    case 'number':
+    case 'bigint': {
+      return true;
+    }
+
+    case 'remotable':
+    case 'copyArray':
+    case 'copyRecord': {
+      // eslint-disable-next-line no-use-before-define
+      return everyPassableChild(passable, isStructure);
+    }
+
+    // Errors are no longer structure
+    case 'error':
+    case 'promise': {
+      return false;
+    }
+    default: {
+      assert.fail(X`Unrecognized passStyle: ${q(passStyle)}`, TypeError);
+    }
+  }
+};
+
+// Like the passStyleCache. An optimization only. Works because comparability
+// is guaranteed stable.
+const structureCache = new WeakMap();
+
+/**
+ * If `passable` is not a Passable, throw. Otherwise say whether it is
```

Comment:
I don't think this throws anymore.  Looks like it returns `true` or `false`.

Reply:
Good catch! Yes.